store/sqlstore/group_supplier.go

```diff
@@ -0,0 +1,864 @@
+	group.UpdateAt = model.GetMillis()
+
+	if err := group.IsValidForUpdate(); err != nil {
+		result.Err = err
+		return result
+	}
+
+	rowsChanged, err := s.GetMaster().Update(group)
+	if err != nil {
+		result.Err = model.NewAppError("SqlGroupStore.GroupUpdate", "store.update_error", nil, err.Error(), http.StatusInternalServerError)
+		return result
+	}
+	if rowsChanged != 1 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupUpdate", "store.sql_group.no_rows_changed", nil, "", http.StatusInternalServerError)
+		return result
+	}
+
+	result.Data = group
+	return result
+}
+
+func (s *SqlSupplier) GroupDelete(ctx context.Context, groupID string, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	if !model.IsValidId(groupID) {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDelete", "model.group.id.app_error", nil, "Id="+groupID, http.StatusBadRequest)
+	}
+
+	var group *model.Group
+	if err := s.GetReplica().SelectOne(&group, "SELECT * from Groups WHERE Id = :Id AND DeleteAt = 0", map[string]interface{}{"Id": groupID}); err != nil {
+		if err == sql.ErrNoRows {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDelete", "store.sql_group.no_rows", nil, "Id="+groupID+", "+err.Error(), http.StatusNotFound)
+		} else {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDelete", "store.select_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		return result
+	}
+
+	time := model.GetMillis()
+	group.DeleteAt = time
+	group.UpdateAt = time
+
+	if rowsChanged, err := s.GetMaster().Update(group); err != nil {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDelete", "store.update_error", nil, err.Error(), http.StatusInternalServerError)
+	} else if rowsChanged != 1 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDelete", "store.sql_group.no_rows_affected", nil, "no record to update", http.StatusInternalServerError)
+	} else {
+		result.Data = group
+	}
+
+	return result
+}
+
+func (s *SqlSupplier) GroupGetMemberUsers(stc context.Context, groupID string, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	var groupMembers []*model.User
+
+	query := `
+		SELECT 
+			Users.* 
+		FROM 
+			GroupMembers 
+			JOIN Users ON Users.Id = GroupMembers.UserId
+		WHERE 
+			GroupMembers.DeleteAt = 0 
+			AND Users.DeleteAt = 0
+			AND GroupId = :GroupId`
+
+	if _, err := s.GetReplica().Select(&groupMembers, query, map[string]interface{}{"GroupId": groupID}); err != nil {
+		if err != sql.ErrNoRows {
+			result.Err = model.NewAppError("SqlGroupStore.GroupGetAllByType", "store.select_error", nil, err.Error(), http.StatusInternalServerError)
+			return result
+		}
+	}
+
+	result.Data = groupMembers
+
+	return result
+}
+
+func (s *SqlSupplier) GroupGetMemberUsersPage(stc context.Context, groupID string, offset int, limit int, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
```

Comment:
Do we need to duplicate `GroupGetMemberUsers` and `GroupGetMemberUsersPage`? Could we just have semantics that a `limit` of `-1` is to select all?

Reply:
I'm not loving the sound of that API with an example invocation of `GetMemberUsersPage(groupID, 0, -1)` to retrieve all members. And the unified method would needa conditional for `order by`, `limit` and `offset`. It doesn't sound cleaner. One could use a struct as a param but that's not consistent with all of the other store/supplier methods. Any other ideas? I could put the first part of the query into a shared constant if code reuse is the main concern.

Reply:
0/5 on my original idea right now, given that our stores aren't exported for third-party usage, and we can always change things later. Let's leave it as-is :)